### PR TITLE
Feature01 auth

### PR DESCRIPTION
**Pull request by thatrafi**
-Adding Auth context to handle auth life cycle 
-Access the redux store and local storage on AuthContext
-AuthContext has local state isLoggedIn which can be used for the children component if they want to know the status of the user authentication

### CODE REVIEW

File: src/routes/PublicRoute.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import React from "react";
```
Comment (mylonasg88) [inline src/routes/PublicRoute.js]: How are you planning to use the Plublic Route?
just `<Route .... />` should be enough if the route is not auth protected.

File: src/routes/PrivateRoute.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+class PrivateRoute extends React.Component {
+    static contextType = AuthContext;
+    // show component only when the user is logged in
+    // Otherwise, redirect the user to /login
+
+    render(){
+        console.log(this.context.isLoggedIn);
+        const { component: Component, ...rest } = this.props;
+        return  <Route {...rest} render={props => (this.context.isLoggedIn ? <Component {...props}/>: <Redirect to={{  pathname: "/login", state: { from: props.location }}}/>)} />
```
Comment (mylonasg88) [inline src/routes/PrivateRoute.js]: It is easier to read if the "logic" is done in "render" function.

Please move the isLoggedIn logic before the "return".

File: src/routes/PublicRoute.js (hunk 1 above)
Comment (thatrafi) [reply src/routes/PublicRoute.js]: I was going to use PublicRoute and Protected Route, i have tried to import and check the isLoggedIn from the context inside ProtectedRoute but somehow it's not getting the last state for isLoggedIn after validate the token, so for now i don't use it instead calling the auth context on App.js and giving the condition inside switch route

Comment (thatrafi) [reply src/routes/PublicRoute.js]: `export const validateByToken = (token) =>{
    return async (dispatch) => {
        // you can dispatch reducer here
        const sendRequest = async () =>{
            const response = await fetch('https://blog-app-react-fb813-default-rtdb.asia-southeast1.firebasedatabase.app/admin.json',
            {
                method : 'GET',
            }
            );
            if(!response.ok){
                throw new Error('Getting admin data failed!')
            }

            const data = response.json()
            return data
        }

        try {
           const responseData =  await sendRequest();
           const adminData = responseData.find( u => u.token === token)
           if(adminData){
                dispatch(authAction.authenticate({message:'Success', isAuthenticated : true, token : adminData.token}))
                console.log('Success validate');
           }else{
                dispatch(authAction.authenticate({message:'Token is not valid', isAuthenticated : false}))
           }
        } catch (error) {
            console.log(error.message);
        }
    }
}`

i have this http async action creators to validate the token and dispatch the state for auth. The auth context, will get the token from local storage and dispatch the validate action on componentDidMount. 

after the state is updated, the auth context also listen the state updates. if isAuthenticated state is true then save the token and set the local state isLoggedIn ( auth-context ) into true

File: src/routes/PrivateRoute.js (hunk 2 above)
Comment (thatrafi) [reply src/routes/PrivateRoute.js]: okay thanks for the suggestion

**Pull request merged**